ctapipe/io/eventfilereader.py

```diff
@@ -148,216 +148,207 @@ def check_file_compatibility(file_path):
-        self._num_events, and return the value.
+        A dictionary containing the metadata of the file. This could include:
+        * is_simulation (bool indicating if the file contains simulated events)
+        * Telescope:Camera names (list if file contains multiple)
+        * Information in the file header
+        * Observation ID
 
         Returns
         -------
-        self._num_events : int
+        dict
         """
+        return self._metadata
 
     @property
-    @abstractmethod
-    def event_id_list(self):
+    def is_stream(self):
         """
-        Abstract property to be defined in child class.
+        Bool indicating if input is a stream. If it is then `__getitem__` and
+        `__len__` are disabled.
 
-        Obtain the number of events from the file, store it as
-        self._event_id_list, and return the value.
+        TODO: Define a method to detect if it is a stream
 
         Returns
         -------
-        self._event_id_list : list[self._num_events]
+        bool
+            If True, then input is a stream.
         """
+        return False
 
     @abstractmethod
-    def read(self, allowed_tels=None, requested_event=None,
-             use_event_id=False):
+    def _generator(self):
         """
         Abstract method to be defined in child class.
 
-        Read the file using the processes required by the readers file-type.
-
-        Parameters
-        ----------
-        allowed_tels : list[int]
-            select only a subset of telescope, if None, all are read. This can
-            be used for example emulate the final CTA data format, where there
-            would be 1 telescope per file (whereas in current monte-carlo,
-            they are all interleaved into one file)
-        requested_event : int
-            Seek to a paricular event index
-        use_event_id : bool
-            If True ,'requested_event' now seeks for a particular event id
-            instead of index
+        Generator where the filling of the `ctapipe.io.containers` occurs.
 
         Returns
         -------
-        source : generator
-            A generator that can be iterated over to obtain events
+        generator
         """
 
-    def get_event(self, requested_event, use_event_id=False):
+    def reset(self):
         """
-        Loop through events until the requested event is found
+        Recreate the generator so it starts from the beginning
+        """
+        self.source = self._generator()
+        self.current_event = None
 
-        Parameters
-        ----------
-        requested_event : int
-            Seek to a paricular event index
-        use_event_id : bool
-            If True ,'requested_event' now seeks for a particular event id
-            instead of index
+    def __iter__(self):
+        # Always reset generator when starting a new iteration
+        self.reset()
+        for event in self.source:
+            self.current_event = event
```

Comment:
We store current event so that:
```
event = reader[5]
event = reader[5]
```
Does not need to start the loop from the beginning to recover the event the second time. This is the only way possible, because the generator has already proceeded to the next event when we call the second line.

Reply:
actually, the way it currently works, the "event" that is returned is a pointer, and there is only ever one instance, so if they do:

```py3
event = next(reader)
reader[5]

event is reader.current_event  
# always True, even if not assigned in the "for event in self.source" loop.
```
`event` will become the 5th event.   This was originally for speed considerations (so we don't have to re-alloc all the arrays and structures each time, rather just copy the data to them).  With this new interface and random-access, however, this may be confusing, so perhaps we should just go for a fully new event each time (with the associated speed reduction).

Reply:
I think we should keep the functionality of event being a pointer, where the data it is pointed to is changed while you iterate through the generator. It is much more efficient, and the situation you have described is no longer a concern for `EventFileReader` as we have removed the `reader[5]` functionality.

Perhaps it should be included in the docstring that the returned event is a pointer, and if one wants to keep that particular event in memory, one should create a `copy.deepcopy` of it.

Reply:
Actually in the current version of EventFileReader (where neither `event = next(reader)` or `reader[5]` is possible), this really isn't a concern.

```python
for event in reader:
   if event.count == 3:
      break
for event2 in reader:
   if event2.count == 5:
      break
assert(event.count == 3)
assert(event2.count == 5)
```

This works as it is a new generator that has been started in the second loop, so new memory has been assigned for the event. 

If we restore `event = next(reader)` functionality we should be careful, and include advice in the docstring, as a confusing situations could arise.

Reply:
>  This was originally for speed considerations (so we don't have to re-alloc all the arrays and structures each time, rather just copy the data to them). With this new interface and random-access, however, this may be confusing, so perhaps we should just go for a fully new event each time (with the associated speed reduction).

How large was the performance impact here?

Reply:
With some quick tests:
Current version (pointer): 80 events per second
Deepcopy the event on yield (dumb): 60 events per second
Create a new DataContainer for each new event: 78 events per second

The relative hit for HessioFileReader is low, as the reader itself is slow. However with TargetIOFileReader I can achieve file reading at around 1200 events per second if I remember correctly.

If it is true that the impact is very small on the efficiency, it certainly seems like something we could consider in the future. But I would advise to do some proper testing first.